### ISSUE TITLE
Pool does not purge LRU, but least recently written

### ISSUE THREAD

**Issue by onnozweers** [dCache/dcache#4853]
Dear dCache devs,

I'm testing a combined tape read+write pool setup (with one pool in the poolgroup) in dCache 5.1 snapshot.

I've noticed that, to make space for new incoming files, the pool does not purge the least recently used file, as I expected; instead, it purges the cached file that has least recently been _written_. I've accessed some files over NFS and over WebDAV, but that does not have any impact: the files are purged in the order they have been written, not the order they have been accessed (for read or write). When reading a file through SRM protocol, the file is pinned for 4 hours, so then it is not purged.

I'm talking about a single pool; so it's about the pool decision, not the PoolManager decision.

Is this behavior intended? I would expect that pool and PoolManager would have the same policy on deciding which files are eligible for purging.

Let me know if you need more information.

Kind regards,
Onno

**Comment by paulmillar** [comment]
I believe we intend to provide (loosely) LRU behaviour.  It's "loosely" because files are deleted when space is needed (typically on upload) and the pool selection does not guarantee that, when accepting an upload, the pool with the oldest cached replica is selected.  One the pool is selected, then I believe we intend to delete replicas that have not been used for the longest.  In other words, LRU.

Therefore I believe this is a bug, perhaps simply that we are not updating the replica's "atime" (stored on the pool, not in the namespace).

**Comment by onnozweers** [comment]
Thanks Paul, I checked the atime of a file and I found that on the first read the atime was adjusted but on the second it wasn't. I then checked the mount properties of the pool partition:

    [root@dolphin12 ~]# mount | grep /space
    /dev/sdb on /space type xfs (rw,noatime,nodiratime,attr2,inode64,noquota)

Possibly the `noatime` is to blame. I'll remove this property and test again.

**Comment by paulmillar** [comment]
IIRC, the replica's atime (when the replica was last read by a dCache user) is actually stored in the filesystem as the file's mtime.  This is because we need to control when it is updated: there are internal operations that should not result in the replica-atime being updated (e.g., being the source for a pool-to-pool transfer).

In addition, the filesystem atime is often a performance issue, so much that filesystems often provide a way of disabling support (e.g., the `noatime` mount option).  Therefore, we cannot rely on this value being maintained by the OS/filesystem.

In summary, the `noatime` mount option shouldn't matter.

**Comment by kofemann** [comment]
Can commit https://github.com/dCache/dcache/commit/0d8df30064af68e5c14d74e6e7d32d64ccb5e8ff be the cause of that behavior?

**Comment by paulmillar** [comment]
@kofemann It's possible, but I don't believe so: the patch didn't modify the value returned by `getLastAccessTime`, which is the value SpaceSweeper2 uses to build the LRU list.

My guess is that either the sweeper's LRU list isn't updated when a file is read, or the replica's file mtime isn't updated.

**Comment by onnozweers** [comment]
Indeed the mtime is not updated:

```
[root@dolphin12 ~]# stat /space/dteam/rwtape/pool/data/0000A1000951004F4DC8BB9ED406CBC64B4E
  File: ‘/space/dteam/rwtape/pool/data/0000A1000951004F4DC8BB9ED406CBC64B4E’
  Size: 1073741824	Blocks: 2097152    IO Block: 4096   regular file
Device: 810h/2064d	Inode: 27920362793  Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2019-05-14 13:46:22.668012514 +0200
Modify: 2019-05-14 11:29:02.683698279 +0200
Change: 2019-05-14 11:29:47.453681949 +0200
 Birth: -
```

The file was written at 11:29, read over WebDAV at 13:46 and also read over WebDAV and over NFS after 13:46.

**Comment by DmitryLitvintsev** [comment]
Not sure if it is related. But at Fermilab (4.2.32) we saw:

```select datestamp, connectiontime, errorcode, errormessage from doorinfo where pnfsid='0000079C6366EDB24ABFBAC0883A36DD781F' order by datestamp desc; ```

         datestamp          | connectiontime | errorcode |                errormessage                
----------------------------+----------------+-----------+--------------------------------------------
 2019-05-16 01:52:18.478-05 |         562411 |         0 | 
 2019-05-15 08:54:33.079-05 |         121523 |         0 | 
 2019-05-15 00:49:28.977-05 |             16 |         5 | Resource temporarily unavailable : Restore
 2019-05-14 01:06:42.486-05 |         394225 |         0 | 
 2019-05-13 01:04:30.745-05 |         609083 |         0 | 


and 

billing=# select datestamp, action  from storageinfo where pnfsid='0000079C6366EDB24ABFBAC0883A36DD781F' order by datestamp desc; 
         datestamp          | action  
----------------------------+---------
 2019-05-15 08:52:31.571-05 | restore
 2019-04-01 00:58:13.752-05 | restore
 2019-02-18 01:04:20.011-06 | restore
 2019-01-08 00:50:39.232-06 | restore
 2018-12-09 01:00:36.918-06 | restore
(5 rows)

File replica was removed on a pool even if the file was accessed every day and there has been plenty of files that are older in terms of access time. 

Interestingly you can see that in terms of "lifetime" determined from replica creation (counting from restore time) the replica existed for about 40 days which corresponds to average replica lifetime on the pool.

**Comment by DmitryLitvintsev** [comment]
Here is what  I observe:

[litvinse@orbis Desktop]$ echo "\s rw-stkendca69a-2@rw-stkendca69a-2Domain sweeper ls -l"  | ssh -p 24223 enstore@fndca  > sweeper-ls.data
Pseudo-terminal will not be allocated because stdin is not a terminal.

take first file:

[litvinse@orbis Desktop]$ head -n 1 sweeper-ls.data
  0 000005080B925F144D99BA66D3225D8DED44  CACHED   3781067824 2019-04-05T14:24:29 2019-04-05T14:24:29 


read it:

[litvinse@orbis Desktop]$ dccp -A pnfs://fndca4a:24125/000005080B925F144D99BA66D3225D8DED44 /dev/null 
3781067824 bytes (3.52 GiB) in 37 seconds (97.5 MiB/s)

check list again:

[litvinse@orbis Desktop]$ echo "\s rw-stkendca69a-2@rw-stkendca69a-2Domain sweeper ls -l"  | ssh -p 24223 enstore@fndca  > sweeper-ls1.data
Pseudo-terminal will not be allocated because stdin is not a terminal.

The file is still first in the list: access time not modified 

[litvinse@orbis Desktop]$ grep '000005080B925F144D99BA66D3225D8DED44' sweeper-ls1.data 
  0 000005080B925F144D99BA66D3225D8DED44  CACHED   3781067824 2019-04-05T14:24:29 2019-04-05T14:24:29 

On the pool partition:

[root@stkendca69a data]# stat 000005080B925F144D99BA66D3225D8DED44
  File: `000005080B925F144D99BA66D3225D8DED44'
  Size: 3781067824	Blocks: 7384904    IO Block: 4096   regular file
Device: 811h/2065d	Inode: 10199573437  Links: 1
Access: (0000/----------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2019-05-16 14:55:58.898727971 -0500
Modify: 2019-04-05 14:24:29.257688000 -0500
Change: 2019-04-05 14:24:29.258005090 -0500

Now, if I do touch 

[root@stkendca69a data]# touch 000005080B925F144D99BA66D3225D8DED44
[root@stkendca69a data]# stat 000005080B925F144D99BA66D3225D8DED44
  File: `000005080B925F144D99BA66D3225D8DED44'
  Size: 3781067824	Blocks: 7384904    IO Block: 4096   regular file
Device: 811h/2065d	Inode: 10199573437  Links: 1
Access: (0000/----------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2019-05-16 15:01:04.451297366 -0500
Modify: 2019-05-16 15:01:04.451297366 -0500
Change: 2019-05-16 15:01:04.451297366 -0500

And re-run dccp again

[litvinse@orbis Desktop]$ dccp -A pnfs://fndca4a:24125/000005080B925F144D99BA66D3225D8DED44 /dev/null 
3781067824 bytes (3.52 GiB) in 36 seconds (100 MiB/s)

[litvinse@orbis Desktop]$ echo "\s rw-stkendca69a-2@rw-stkendca69a-2Domain sweeper ls -l"  | ssh -p 24223 enstore@fndca  > sweeper-ls2.data
Pseudo-terminal will not be allocated because stdin is not a terminal.
[litvinse@orbis Desktop]$ grep '000005080B925F144D99BA66D3225D8DED44' sweeper-ls2.data 
  0 000005080B925F144D99BA66D3225D8DED44  CACHED   3781067824 2019-04-05T14:24:29 2019-04-05T14:24:29 

No update. 

So yeah. Looks like Sweeper based on last access LRU is broken. This is a very serious issue. 

(The reason I looked at this issue - we have number of tickets opened about files disappearing from 
cache unexpectedly)

**Comment by DmitryLitvintsev** [comment]
@onnozweers I think the problem is "last access time is not updated on file access" (fact that this means 
that files get removed based on creation time follows from that)

**Comment by DmitryLitvintsev** [comment]
git revert 0d8df30064af68e5c14d74e6e7d32d64ccb5e8ff 
rebuilt, re-installed on a pool on a test system:

```[dmsdca06] (PoolManager@dCacheDomain) enstore > \c r-dmsdca02-1
[dmsdca06] (r-dmsdca02-1@r-dmsdca02-1Domain) enstore > sweeper ls -l 
  0 0000BA83AD6C2B1E4B0F9EC565C09812E77D  CACHED  42949672960 2019-05-16T15:24:12 2018-12-07T22:31:45 
  1 000047BFEC84AFFB46C0860B2133FAFAC042  CACHED    179979217 2019-05-16T15:24:11 2018-12-11T21:38:27 
  2 00005913ED973796408FBA81438595F6BCE3  CACHED    179871155 2019-05-16T15:24:12 2018-12-13T16:56:25 
  3 000022CA2DCD05DE4048997996C3EF3F13FE  CACHED    179613617 2019-05-16T15:24:12 2018-12-14T08:54:27 
  4 000033472620EF2B4162B7F60ACA9AED2817  CACHED    180032132 2019-05-16T15:24:11 2018-12-14T09:10:49 
  5 0000B04D5294201E46ECAE42434B48019779  CACHED    179442777 2019-05-16T15:24:12 2018-12-14T09:30:53 
  6 00007758521D18314A4291217E9627E25364  CACHED    179512608 2019-05-16T15:24:12 2018-12-14T13:01:09 
  7 00009FD10113696E41D4A68BB8FAD63CD6D8  CACHED     35983535 2019-05-16T15:24:12 2019-01-24T13:08:27 
  8 0000B3611818A7C84CCE9C95216551590180  CACHED      2265398 2019-05-16T15:24:11 2019-01-24T13:19:56 
  9 00003D97C7EFF9AA4C69B20057C851713201  CACHED     34999201 2019-05-16T15:24:11 2019-01-25T13:12:21 
 10 00001622ACAC2D684F629F1C2D3F197842D1  CACHED     35698348 2019-05-16T15:24:11 2019-01-25T13:13:48 
 11 000053791EFAD0C4423486947B77BE24D74E  CACHED     27891148 2019-05-16T15:24:12 2019-01-25T13:18:39 
 12 00008295D45E864E479899B915320577FF4B  CACHED     32728734 2019-05-16T15:24:12 2019-01-25T13:57:23 
 13 00004798CC4F6CAF48869CA3F436D6214AB0  CACHED     21096041 2019-05-16T15:24:11 2019-01-25T15:17:37 
 14 0000526828849A8E40C48830783EC9E068CA  CACHED   2019257023 2019-05-16T15:24:12 2019-01-26T18:12:41 
 15 0000546D12324B3942C8B308DBF5E2E4680E  CACHED     86619108 2019-05-16T15:24:12 2019-04-02T14:55:43 
 16 0000B9CDE8B1407C4FC690FF0D9A2934EFAA  CACHED    121429456 2019-05-16T15:24:11 2019-04-17T09:44:25 
 17 0000CCD90C3C53ED474888192C663B9E45C8  CACHED     58320304 2019-05-16T15:24:11 2019-04-17T09:52:10 
 18 0000A2A3BD8828A741B8BA8763BB26E59C9D  CACHED       801528 2019-05-16T15:24:12 2019-04-30T09:21:42 

[dmsdca06] (r-dmsdca02-1@r-dmsdca02-1Domain) enstore > 
```

[litvinse@orbis Desktop]$ dccp -A pnfs://dmsdca07.fnal.gov:24125/000047BFEC84AFFB46C0860B2133FAFAC042 /dev/null 
179979217 bytes (172 MiB) in 3 seconds (57.2 MiB/s)

(second file in sweeper ls output) 

the output is unchanged after copy:

```[dmsdca06] (r-dmsdca02-1@r-dmsdca02-1Domain) enstore > sweeper ls -l 
  0 0000BA83AD6C2B1E4B0F9EC565C09812E77D  CACHED  42949672960 2019-05-16T15:24:12 2018-12-07T22:31:45 
  1 000047BFEC84AFFB46C0860B2133FAFAC042  CACHED    179979217 2019-05-16T15:24:11 2018-12-11T21:38:27 
  2 00005913ED973796408FBA81438595F6BCE3  CACHED    179871155 2019-05-16T15:24:12 2018-12-13T16:56:25 
  3 000022CA2DCD05DE4048997996C3EF3F13FE  CACHED    179613617 2019-05-16T15:24:12 2018-12-14T08:54:27 
  4 000033472620EF2B4162B7F60ACA9AED2817  CACHED    180032132 2019-05-16T15:24:11 2018-12-14T09:10:49 
  5 0000B04D5294201E46ECAE42434B48019779  CACHED    179442777 2019-05-16T15:24:12 2018-12-14T09:30:53 
  6 00007758521D18314A4291217E9627E25364  CACHED    179512608 2019-05-16T15:24:12 2018-12-14T13:01:09 
  7 00009FD10113696E41D4A68BB8FAD63CD6D8  CACHED     35983535 2019-05-16T15:24:12 2019-01-24T13:08:27 
  8 0000B3611818A7C84CCE9C95216551590180  CACHED      2265398 2019-05-16T15:24:11 2019-01-24T13:19:56 
  9 00003D97C7EFF9AA4C69B20057C851713201  CACHED     34999201 2019-05-16T15:24:11 2019-01-25T13:12:21 
 10 00001622ACAC2D684F629F1C2D3F197842D1  CACHED     35698348 2019-05-16T15:24:11 2019-01-25T13:13:48 
 11 000053791EFAD0C4423486947B77BE24D74E  CACHED     27891148 2019-05-16T15:24:12 2019-01-25T13:18:39 
 12 00008295D45E864E479899B915320577FF4B  CACHED     32728734 2019-05-16T15:24:12 2019-01-25T13:57:23 
 13 00004798CC4F6CAF48869CA3F436D6214AB0  CACHED     21096041 2019-05-16T15:24:11 2019-01-25T15:17:37 
 14 0000526828849A8E40C48830783EC9E068CA  CACHED   2019257023 2019-05-16T15:24:12 2019-01-26T18:12:41 
 15 0000546D12324B3942C8B308DBF5E2E4680E  CACHED     86619108 2019-05-16T15:24:12 2019-04-02T14:55:43 
 16 0000B9CDE8B1407C4FC690FF0D9A2934EFAA  CACHED    121429456 2019-05-16T15:24:11 2019-04-17T09:44:25 
 17 0000CCD90C3C53ED474888192C663B9E45C8  CACHED     58320304 2019-05-16T15:24:11 2019-04-17T09:52:10 
 18 0000A2A3BD8828A741B8BA8763BB26E59C9D  CACHED       801528 2019-05-16T15:24:12 2019-04-30T09:21:42 

```

**Comment by DmitryLitvintsev** [comment]
deployed 4.2.0 - same behavior 
deployed 3.2.53 - correct behavior 

```[dmsdca06] (r-dmsdca02-1@r-dmsdca02-1Domain) enstore > sweeper ls -l 
  0 0000BA83AD6C2B1E4B0F9EC565C09812E77D  CACHED  42949672960 2019-05-16T15:46:37 2018-12-07T22:31:45 
  1 000047BFEC84AFFB46C0860B2133FAFAC042  CACHED    179979217 2019-05-16T15:46:37 2019-05-16T15:46:56  <---- changed 
  2 00005913ED973796408FBA81438595F6BCE3  CACHED    179871155 2019-05-16T15:46:37 2018-12-13T16:56:25 
  3 000022CA2DCD05DE4048997996C3EF3F13FE  CACHED    179613617 2019-05-16T15:46:37 2018-12-14T08:54:27 
  4 000033472620EF2B4162B7F60ACA9AED2817  CACHED    180032132 2019-05-16T15:46:37 2018-12-14T09:10:49 
  5 0000B04D5294201E46ECAE42434B48019779  CACHED    179442777 2019-05-16T15:46:37 2018-12-14T09:30:53 
  6 00007758521D18314A4291217E9627E25364  CACHED    179512608 2019-05-16T15:46:37 2018-12-14T13:01:09 
  7 00009FD10113696E41D4A68BB8FAD63CD6D8  CACHED     35983535 2019-05-16T15:46:37 2019-01-24T13:08:27 
  8 0000B3611818A7C84CCE9C95216551590180  CACHED      2265398 2019-05-16T15:46:37 2019-01-24T13:19:56 
  9 00003D97C7EFF9AA4C69B20057C851713201  CACHED     34999201 2019-05-16T15:46:37 2019-01-25T13:12:21 
 10 00001622ACAC2D684F629F1C2D3F197842D1  CACHED     35698348 2019-05-16T15:46:37 2019-01-25T13:13:48 
 11 000053791EFAD0C4423486947B77BE24D74E  CACHED     27891148 2019-05-16T15:46:37 2019-01-25T13:18:39 
 12 00008295D45E864E479899B915320577FF4B  CACHED     32728734 2019-05-16T15:46:37 2019-01-25T13:57:23 
 13 00004798CC4F6CAF48869CA3F436D6214AB0  CACHED     21096041 2019-05-16T15:46:37 2019-01-25T15:17:37 
 14 0000526828849A8E40C48830783EC9E068CA  CACHED   2019257023 2019-05-16T15:46:37 2019-01-26T18:12:41 
 15 0000546D12324B3942C8B308DBF5E2E4680E  CACHED     86619108 2019-05-16T15:46:37 2019-04-02T14:55:43 
 16 0000B9CDE8B1407C4FC690FF0D9A2934EFAA  CACHED    121429456 2019-05-16T15:46:37 2019-04-17T09:44:25 
 17 0000CCD90C3C53ED474888192C663B9E45C8  CACHED     58320304 2019-05-16T15:46:37 2019-04-17T09:52:10 
 18 0000A2A3BD8828A741B8BA8763BB26E59C9D  CACHED       801528 2019-05-16T15:46:37 2019-04-30T09:21:42 

```

**Comment by DmitryLitvintsev** [comment]
Here we go:

[litvinse@orbis dcache]$ git bisect good 
1d2f7ee1908afc77e8f068ea6b245ca93b3f944d is the first bad commit
commit 1d2f7ee1908afc77e8f068ea6b245ca93b3f944d
Author: Paul Millar <paul.millar@desy.de>
Date:   Thu Dec 7 20:59:01 2017 +0100

    pool: add IO statistics to monitor storage performance
    
    Motivation:
    
    First, make IoStatistics generally available so that information about
    the underlying storage system performance is available.  This has the
    potential for providing useful insight into poor performance: whether
    the problem is with the client/network delivering data slowly, or the
    poor performance of the underlying storage device.  This monitoring may
    also allow a more adaptive behaviour, such transfers based on
    performance of the underlying performance.
    
    Second, to demonstrate a generic idea on how new pool features may be
    introduced cleanly, by allowing a RepositoryChannel to be optionally
    cast to something implementation specific.
    
    Modification:
    
    Forwarding classes (ones that delegate operations to some other
    instance) for ReplicaRecord and ReplicaStore are introduced.
    
    Introduce specialist ReplicaStore and ReplicaRecord classes (based on
    these forwarding classes) that allow optional injection of
    IoStatisticsChannel: channels opened with the ENABLE_IO_STATISTICS open
    option.  Without this flag, statistics are not gathered.
    
    Finally, a handful of metrics are added to billing log files, both read
    and write operations. These metrics are the mean instantaneous bandwidth
    (data transferred / time taken, evaluated for each IO request), the
    duration spent with at least one pending IO request, and the duration
    spend with no pending IO request.
    
    Result:
    
    Billing can now log, for read and write operations, the mean bandwidth
    of each IO request and time that the disk system was handling requests
    and idling.  NB. the dCache configuration that describes the billing log
    format must be updated before these values are actually recorded.
    
    Target: master
    Require-notes: yes
    Require-book: yes
    Patch: https://rb.dcache.org/r/10684/
    Acked-by: Tigran Mkrtchyan

:040000 040000 2ee538aa1392e2402ff67672620a2176f5ffaebd 97b8bfa574db2769c16f78ffc3b497035b3bc9c4 M	modules
:040000 040000 b3b0da99131a68cd23309540831d520106cdb36c 5f0dfa7839e156d8d6a223a5547e0355fb8cfef2 M	skel
[litvinse@orbis dcache]$

**Comment by paulmillar** [comment]
There are actually two regressions with LRU: the first is obvious, the second more subtle.

The more obvious one that @DmitryLitvintsev pointed out (regression with commit 1d2f7ee) is fixed with https://rb.dcache.org/r/11745/

The more subtle one (regression with commit 9830f2d) results in a file access not updating the LRU ordering, despite the last-access-time changing.  This is fixed with https://rb.dcache.org/r/11746/

**Comment by onnozweers** [comment]
Thanks Paul and Dmitry!
